src/client/opamSwitchCommand.ml

```diff
@@ -492,29 +505,69 @@ let read_overlays (read: package -> OpamFile.OPAM.t option) packages =
     packages
     OpamPackage.Name.Map.empty
 
-let export ?(full=false) filename =
+let read_extra_files ~repos_roots (read: package -> OpamFile.OPAM.t option) data packages =
+  OpamPackage.Set.fold (fun nv acc ->
+      match read nv with
+      | None -> acc
+      | Some opam ->
+        match OpamFile.OPAM.get_extra_files ~repos_roots opam with
+        | [] -> acc
+        | files ->
+          (* read file and embed them into opam *)
+          List.fold_left (fun acc (file, _rel_file, hash) ->
+              if OpamFilename.exists file &&
+                 OpamHash.check_file (OpamFilename.to_string file) hash then
+                let value = OpamFilename.read file in
+                let value' = Base64.encode_string value in
+                OpamHash.Map.add hash value' acc
+              else begin
+                OpamConsole.warning "Ignoring file %s with invalid hash"
+                  (OpamFilename.to_string file);
+                acc
+              end)
+            acc files)
+    packages data
+
+let export rt ?(full=false) ?(freeze = false) filename =
   let switch = OpamStateConfig.get_switch () in
   let root = OpamStateConfig.(!r.root_dir) in
   let export =
     OpamFilename.with_flock `Lock_none (OpamPath.Switch.lock root switch)
     @@ fun _ ->
     let selections = S.safe_read (OpamPath.Switch.selections root switch) in
     let overlays =
-      read_overlays (fun nv ->
-          OpamFileTools.read_opam
-            (OpamPath.Switch.Overlay.package root switch nv.name))
-        selections.sel_pinned
+        read_overlays (not freeze) (fun nv ->
+            OpamFileTools.read_opam
+              (OpamPath.Switch.Overlay.package root switch nv.name))
+          selections.sel_pinned
     in
     let overlays =
-      if full then
+      if full || freeze then
         OpamPackage.Name.Map.union (fun a _ -> a) overlays @@
-        read_overlays (fun nv ->
+        read_overlays (not freeze) (fun nv ->
             OpamFile.OPAM.read_opt
               (OpamPath.Switch.installed_opam root switch nv))
           (selections.sel_installed -- selections.sel_pinned)
-      else overlays
+      else
+        overlays
+    in
+    let extra_files =
+      if freeze then
+        let repos_roots = OpamRepositoryState.get_root rt in
+        let data =
+          read_extra_files ~repos_roots (fun nv ->
+              OpamFileTools.read_opam
```

Comment:
It is possible to factorise read overlay opam function and read installed opam one. each one is used two times.

Reply:
It can even be the opposite: have two function `on_overlay_pinned_opam` and `on_installed_opam` that takes the function to call, e.g? `on_installed_opam @@ read_extra_files ~repos_roots`.

Reply:
_just a suggestion/thinking about it, i'd just be happy with the factorisation :)_

Reply:
done in 4633e78e